source/common/http/header_map_impl.h

```diff
@@ -256,6 +291,176 @@ class HeaderMapImpl : NonCopyable {
   uint64_t cached_byte_size_ = 0;
 };
 
+#elif HEADER_MAP_TYPE == HEADER_MAP_TYPE_FLAT_HASH_MAP
+
+/**
+ * Implementation of Http::HeaderMap that is based on absl::flat_hash_map.
+ */
+class HeaderMapImpl : public virtual HeaderMap, NonCopyable {
+public:
+  static void copyFrom(HeaderMap& lhs, const HeaderMap& rhs);
+  // The value_type of iterator must be pair, and the first value of them must be LowerCaseString.
+  // If not, it won't be compiled successfully.
+  template <class It> static void initFromInitList(HeaderMap& new_header_map, It begin, It end) {
+    for (auto it = begin; it != end; ++it) {
+      static_assert(std::is_same<decltype(it->first), LowerCaseString>::value,
+                    "iterator must be pair and the first value of them must be LowerCaseString");
+      HeaderString key_string;
+      key_string.setCopy(it->first.get().c_str(), it->first.get().size());
+      HeaderString value_string;
+      value_string.setCopy(it->second.c_str(), it->second.size());
+      new_header_map.addViaMove(std::move(key_string), std::move(value_string));
+    }
+  }
+
+  // Performs a manual byte size count for test verification.
+  void verifyByteSizeInternalForTest() const;
+
+  // Http::HeaderMap
+  bool operator==(const HeaderMap& rhs) const override;
+  bool operator!=(const HeaderMap& rhs) const override;
+  void addViaMove(HeaderString&& key, HeaderString&& value) override;
+  void addReference(const LowerCaseString& key, absl::string_view value) override;
+  void addReferenceKey(const LowerCaseString& key, uint64_t value) override;
+  void addReferenceKey(const LowerCaseString& key, absl::string_view value) override;
+  void addCopy(const LowerCaseString& key, uint64_t value) override;
+  void addCopy(const LowerCaseString& key, absl::string_view value) override;
+  void appendCopy(const LowerCaseString& key, absl::string_view value) override;
+  void setReference(const LowerCaseString& key, absl::string_view value) override;
+  void setReferenceKey(const LowerCaseString& key, absl::string_view value) override;
+  void setCopy(const LowerCaseString& key, absl::string_view value) override;
+  uint64_t byteSize() const override;
+  const HeaderEntry* get(const LowerCaseString& key) const override;
+  void iterate(ConstIterateCb cb, void* context) const override;
+  void iterateReverse(ConstIterateCb cb, void* context) const override;
+  Lookup lookup(const LowerCaseString& key, const HeaderEntry** entry) const override;
+  void clear() override;
+  size_t remove(const LowerCaseString& key) override;
+  size_t removePrefix(const LowerCaseString& key) override;
+  size_t size() const override { return all_headers_num_; }
+  bool empty() const override { return all_headers_num_ == 0; }
+  void dumpState(std::ostream& os, int indent_level = 0) const override;
+  bool canCoalesce(const HeaderString& key) const {
+    // At the moment only inline-headers are coalesced.
```

Comment:
FWIW, in a real implementation, to match existing behavior, you'd need to coalesce the O(1) headers.

Reply:
Not sure I understand what you are referring to.
All the O(1) non-custom headers are being coalesced.
Each concrete header-map (e.g., `ResponseHeaderMapImpl`) has a set of header names (those defined by the macros in [header_map.h](https://github.com/envoyproxy/envoy/blob/f634d83a1a58a44f3a44ebb67888660577354d9f/include/envoy/http/header_map.h#L278)). Upon the first access of such header, it is instantiated as a "coalesced header".

It is possible to add the same functionality for custom inline registered headers. Although if it is needed, then I'd advocate for a better API, that clearly defines that a newly registered header can be coalesced or not (and probably also if it should only hold numerical values).

Reply:
Ack, I think the terminology is what confused me; I tend to think of inlined headers as just short headers that haven't been expanded out to independent data structures, maybe say explicitly that this is O(1).